src/fetchers/tarball-fetcher.js

```diff
@@ -82,13 +82,30 @@ export default class TarballFetcher extends BaseFetcher {
         const expectHash = this.hash;
         const actualHash = validateStream.getHash();
         if (!expectHash || expectHash === actualHash) {
+          const tarballMirrorPath = this.getTarballMirrorPath();
+          const tarballCachePath = this.getTarballCachePath();
+
+          if (tarballMirrorPath) {
+            validateStream.pipe(fs.createWriteStream(tarballMirrorPath)).on('error', reject);
+          }
+
+          if (tarballCachePath) {
+            validateStream.pipe(fs.createWriteStream(tarballCachePath)).on('error', reject);
+          }
```

Comment:
Agree with @arcanis. I think what we need is keeping this in the old place and just deleting the file on validation failure before we reject the promise. This would also prevent going over the file two times.